### PR TITLE
Add Equilibrium profiling of global and final optimization

### PR DESCRIPTION
**Pull request by nouiz**

### CODE REVIEW

**Comment by lamblin** [comment]
This PR needs a rebase.

**Comment by nouiz** [comment]
I rebased this PR.

File: theano/sandbox/cuda/cuda_ndarray.cu [hunk 1]
```diff
@@ -1064,9 +1064,12 @@ CudaNdarray_TakeFrom(CudaNdarray * self, PyObject *args){
...
     } else {
-        PyErr_SetString(PyExc_TypeError,
-                        "CudaNdarray_TakeFrom: need a CudaNdarray(float32) that"
-                        " is a view from int64 data for indices");
+        PyObject* py_s = PyObject_Str(indices_obj);
+        const char* s = PyString_AsString(py_s);
+        PyErr_Format(PyExc_TypeError,
+                     "CudaNdarray_TakeFrom: need an ndarray of int64 or a"
+                     " CudaNdarray(float32) that is a view from int64 data"
+                     " for indices. Got %s", s);
```
Comment (lamblin) [inline theano/sandbox/cuda/cuda_ndarray.cu]: I think `py_s` should be decrefed before returning, since `PyObject_Str` returns a fresh reference.

**Comment by lamblin** [comment]
Apart from that, I think it makes sense.

**Comment by nouiz** [comment]
I fixed the comment.

**Pull request merged**